### PR TITLE
Top Down Prediction Errors for Latent State Update 

### PR DESCRIPTION
**Pull request by Nardos24**
This PR updates the training dynamics to follow predictive coding principles more closely.

**Error propagation mechanism:**

- Each layer now caches its prediction error and passes it down to the lower layer. [6e6fa7e](https://github.com/iCog-Labs-Dev/PC-Transformers/commit/6e6fa7e2e27826e134aab0a2a8d80547b68284bb)

- The lower layer then incorporates this error into its latent state updates, enabling bidirectional refinement (bottom-up correction and top-down constraint). [e2e2821](https://github.com/iCog-Labs-Dev/PC-Transformers/commit/e2e28211f0050574235b2e19f95d7e29cfe79531)

 

- This creates a hierarchical feedback loop where both local prediction errors and higher-layer signals contribute to updating latent states.

**Predictive Coding energy metric:**

- Replaced the previous MSE-based energy calculation with predictive coding (PC) energy. [006edab](https://github.com/iCog-Labs-Dev/PC-Transformers/commit/006edab61d07644f94a8e2bd284b807c2db5509d)

- PC energy is computed layerwise from the prediction errors. [adcd365](https://github.com/iCog-Labs-Dev/PC-Transformers/commit/adcd3656c4076509ce9cd4f725ff949f072e0baa)  

**Remove reassigning nn parameters**  

   **previous implementation** 
```word_weight = word_layer.weight.data.index_add(0, flat_input_ids, local_lr * flat_update)
word_layer.weight = nn.Parameter(word_weight)
````

- Each update created a brand-new parameter object and reassigned it to the layer. This was unnecessary and could lead to extra memory allocations.

**current implementation** 

```word_layer.weight.data.index_add_(0, flat_input_ids, local_lr * flat_update)```

- Updates the parameter values directly in-place, without replacing the parameter object. This is more efficient and keeps the model structure stable. [5ff485e](https://github.com/iCog-Labs-Dev/PC-Transformers/pull/91/commits/5ff485e5f689966f3d23bd88a54b816e5562455a)

**clamping values** 
To prevent NaN energy and infinite perplexity caused by including the top-down error, I applied weight clamping methods to stabilize the updates.[d3a5057](https://github.com/iCog-Labs-Dev/PC-Transformers/commit/d3a50571ae7b709b31253b418c3b2a6d7481fefa)

**Note:**

-  Use predictive coding (PC) energy for the output layer. 
- Parameter sizes should be adjusted according to the dataset: for smaller datasets like PTB, use smaller parameter values, and for larger datasets like OpenWebText, use larger parameters to ensure stable training.

**Update Rule**

Latent states are updated using the predictive coding update formula:

``` X_new=X_old+𝜂⋅(𝑊⋅bu_err−td_err) ```

where:

𝜂 = learning rate for state updates

W⋅bu_err = bottom-up error correction projected through the weight matrix

td_err = top-down prediction error from higher layers

### CODE REVIEW

**Review by quinbez — COMMENTED** [review]
I encountered an issue when training the model. Even with our latest tuned hyperparameters, the problem persists. Please fix this and ensure the implementation is tested before submitting a PR.

Current output:
```
12.51M parameters  
Epoch 1/20  
  Batch 10/161392 | Energy: nan | Perplexity: inf  
  Batch 20/161392 | Energy: nan | Perplexity: inf  
  Batch 30/161392 | Energy: nan | Perplexity: inf  
  ...
  ```
